metals/src/main/scala/scala/meta/internal/metals/Compilers.scala

```diff
@@ -0,0 +1,145 @@
+package scala.meta.internal.metals
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+import ch.epfl.scala.bsp4j.ScalaBuildTarget
+import ch.epfl.scala.bsp4j.ScalacOptionsItem
+import java.util.Collections
+import org.eclipse.lsp4j.CompletionItem
+import org.eclipse.lsp4j.CompletionList
+import org.eclipse.lsp4j.CompletionParams
+import org.eclipse.lsp4j.Hover
+import org.eclipse.lsp4j.SignatureHelp
+import org.eclipse.lsp4j.TextDocumentPositionParams
+import org.eclipse.lsp4j.jsonrpc.CancelChecker
+import scala.concurrent.ExecutionContext
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.pc.ScalaPresentationCompiler
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.PresentationCompiler
+import scala.meta.pc.SymbolSearch
+import scala.tools.nsc.Properties
+
+/**
+ * Manages lifecycle for presentation compilers in all build targets.
+ *
+ * We need a custom presentation compiler for each build target since
+ * build targets can have different classpaths and compiler settings.
+ */
+class Compilers(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+    search: SymbolSearch,
+    embedded: Embedded,
+    statusBar: StatusBar
+)(implicit ec: ExecutionContext)
+    extends Cancelable {
+
+  // Not a TrieMap because we want to avoid loading duplicate compilers for the same build target.
+  val jcache = Collections.synchronizedMap(
```

Comment:
But why a synchronized version vs `java.util.concurrent.ConcurrentHashMap`?

Reply:
`j.u.c.ConcurrentHashMap` can deadlock in `computeIfAbsent` when the compute function is expensive